CHANGELOG.md

```diff
@@ -1,10 +1,11 @@
 # Release History
 
-## 0.18.1 (unreleased)
+## 0.18.1 (Unreleased)
 
 ### Other Changes
 
 * The connection mux goroutine has been removed, eliminating a potential source of deadlocks.
+* Clarified docs that all received messages must be acknowledged.
```

Comment:
Looking at this some more I think this isn't always true. If the sender that corresponds to our receiver says it's sender-settle-mode is 'settled' then there's nothing for us to ack on the client side. The messages are already pre-settled.

In that case I don't think you'd need or want to settle messages.

Reply:
You bring up a good point.  If the sender is configured as `SenderSettleModeSettled` then there's no need to settle the message (I will fix the docs to include this).  In the other cases, we don't at present export the `settled` state on the message, so callers can't tell if the message is settled (we check for this internally in our settlement APIs to see if we need to send a disposition frame).  So the question becomes, do we want to export the `settled` state, or just say "always call a settlement API and the right thing happens".

Reply:
Since 'settled' is part of the AMQP transfer frame (ie, per message), it makes sense that it'd be included in the returned amqp.Message{}.

Reading the spec more, I think there are cases where the settled state is not known ahead of time or might be different per message:

1. The AMQP broker "ignores" the receiver's suggested sender mode (apparently it's allowed to, but it's recommended to not DO that). So they might come back unsettled or settled, or at least, there's no way for the client to assume that it came back one way or another.
2. The user chooses "Mixed", which (I assume) means that messages aren't universally settled or unsettled.

So I'd vote per-message flag (ie, Settled bool) and then that's probably good enough. We could still no-op if they call, or maybe even just error out so the user isn't expecting to be able to "reverse" the settlement if it's already been settled.

Reply:
For your first point, while the peer is allowed to reject the request, our current implementation treats this as a hard error and will detach the link.  In this respect, the peer will never be configured in a different mode than what was requested (we see this with brokers like ActiveMQ, which doesn't support RSMModeSecond).

I don't think we should return an error when attempting to settle a message that's already been settled as that just complicates the logic (leave it a no-op as it is now).

Reply:
I've opened https://github.com/Azure/go-amqp/issues/215 to track this.